### ISSUE TITLE
New player causes: duplicate key value violates unique constraint "auth_name_key"

### ISSUE THREAD

**Issue by Bastrabun** [minetest/minetest#11877]
##### Minetest version
```
mt 5.4.1
```

##### OS / Hardware

Operating system: debian 11
CPU: AMD Epyc 7282
Postgres postgresql-13  13.5-0+deb11u1

##### Summary
```
2021-12-19 11:56:00: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod '' in callback createAuth(): PostgreSQL database error: ERROR: duplicate key value violates unique constraint "auth_name_key"
2021-12-19 11:56:00: ERROR[Main]: DETAIL: Key (name)=(basu) already exists.
2021-12-19 11:56:00: ERROR[Main]: stack traceback:
2021-12-19 11:56:00: ERROR[Main]: [C]: at 0x557f09c49420
```
Backup of 2021-12-16-19-29 does not have a player called "basu" in the auth table, now there is one but logindate is -1

There is a different playername that includes "basu" , like Xyzbasu

##### Repro

I do not have a repro of any kind, the error happened over the course of 2 years more than once, while the other 16 000 accs had no trouble of that kind.

For my reference 1189

**Comment by fluxionary** [comment]
This is triggered when someone deliberately floods a server w/ identical logins. Someone is currently using this to DOS multiple servers.

**Comment by fluxionary** [comment]
Steps to replicate:

* start a local instance of 5.5 against a postgres db (i'm on 13.6)
* set up this command line client program: https://github.com/minetest-go/minetest_client.git
* run the following script (you may need to tweak the sleep time, too slow and no fatal error, too quick and you get a different (less useful) error message)
```bash
#!/usr/bin/env bash

exitfn() {
    trap - INT
    killall -9 minetest_client
    exit
}

trap "exitfn" INT

word=`shuf -n 1 /usr/share/dict/words|sed 's/[^a-zA-Z]//g'`
for i in {1..100}
    do
    name="$word$i"
    echo "$name"
    for j in {1..5}
    do
        ./minetest_client -username "$name" -host 127.0.0.1 >/dev/null 2>/dev/null &
        sleep 0.003
    done
    sleep 5
    killall -9 minetest_client
done

trap - INT
```

**Comment by fluxionary** [comment]
This is fundamentally a race condition - the check for whether authentication data exists doesn't happen synchronously w/ the attempt to set new authentication data.

**Comment by fluxionary** [comment]
Problem is, even after adding a failure mode for that statement, I still get another error:

```
AsyncErr: Lua: Runtime error from mod '' in callback getAuth(): PostgreSQL database error: ERROR:  current transaction is aborted, commands ignored until end of transaction block

stack traceback:
	[C]: in function 'read'
	/usr/share/minetest/builtin/game/auth.lua:14: in function </usr/share/minetest/builtin/game/auth.lua:12>
```

Which is incredibly opaque to me

**Comment by fluxionary** [comment]
It's being triggered at the following line:

https://github.com/minetest/minetest/blob/b55d7cd45a304a8a5b8337a3c61defb6e5c27dd9/src/database/database-postgresql.cpp#L691

I still have no idea *why*, and have maybe run up against a wall w.r.t. my understanding of the game architecture. I can't see anywhere that a transaction isn't properly terminated. Is it possible that a connection with a failed transaction is getting reused by another thread before it can rollback?

**Comment by fluxionary** [comment]
> Is it possible that a connection with a failed transaction is getting reused by another thread before it can rollback?

Pretty certain that is what is going on, given the existence of something called "connectionthreads". Probably the solution there is to create proper synchronization locks around access to the DB connection. My personal experience w/ c++ synch is pretty minimal, but I'll give it a shot unless someone else wants to take over.

**Comment by BuckarooBanzay** [comment]
fyi: this is the pendant for the `sqlite` backend: https://github.com/minetest/minetest/issues/9498

**Comment by appgurueu** [comment]
> fyi: this is the pendant for the `sqlite` backend: #9498

SQLite doesn't need "synchronization around DB connection access" (fluxionary) as it has a connection lock anyways.

**Comment by Oblomov** [comment]
@fluxionary could you set up a trace on the db or in the server code to see all the db commands that get called, in sequence, possibly with timing informaiton? With your repro, it should be possible to identify the sequence that leads to the issue and this might help identify the source of the issue.

**Comment by Oblomov** [comment]
@fluxionary to wit, something like the following patch applied to the server:
````
diff --git a/src/database/database-postgresql.h b/src/database/database-postgresql.h
index 81b4a2b10..88dc392d2 100644
--- a/src/database/database-postgresql.h
+++ b/src/database/database-postgresql.h
@@ -24,6 +24,12 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "database.h"
 #include "util/basic_macros.h"
 
+#define DEBUG_DB 1
+#if DEBUG_DB
+#include "debug.h"
+#include <chrono>
+#endif
+
 class Settings;
 
 class Database_PostgreSQL: public Database
@@ -70,6 +76,10 @@ class Database_PostgreSQL: public Database
 		const int *paramsLengths = NULL, const int *paramsFormats = NULL,
 		bool clear = true, bool nobinary = true)
 	{
+#if DEBUG_DB5
+		warningstream << std::put_time("%Y-%m-%d %H:%M;%S", std::chrono::system_clock::now()) << "\t" << m_conn << std::endl;
+#endif
+
 		return checkResults(PQexecPrepared(m_conn, stmtName, paramsNumber,
 			(const char* const*) params, paramsLengths, paramsFormats,
 			nobinary ? 1 : 0), clear);
````
might help with the tracing.

**Comment by sfan5** [comment]
To debug this it's not so important what the statements are, it's important to have a backtrace of the conflicting db calls. Because all packet handling and server steps happen on the same thread it should never be possible for two calls to conflict.

If you compile Minetest with debug info and plain Lua **gdb** can probably give you the backtrace you need. It's possible to set it to automatically retrieve a backtrace when hitting a breakpoint:
<pre>
b database-postgresql.cpp:691
commands
bt full
c
end
</pre>

**Comment by BuckarooBanzay** [comment]
> Because all packet handling and server steps happen on the same thread it should never be possible for two calls to conflict.

Afaik: every connection (client) has its own thread.

In one connection you could enter here: `Server::handleCommand_FirstSrp`
https://github.com/minetest/minetest/blob/480d5f2d51ca8f7c4400b0918bb53b776e4ff440/src/network/serverpackethandler.cpp#L1500

And in another at the same time here: `Server::handleCommand_SrpBytesM`
https://github.com/minetest/minetest/blob/480d5f2d51ca8f7c4400b0918bb53b776e4ff440/src/network/serverpackethandler.cpp#L1698

Both will execute the following method concurrently:
https://github.com/minetest/minetest/blob/480d5f2d51ca8f7c4400b0918bb53b776e4ff440/src/script/cpp_api/s_server.cpp#L106-L121

The script environment will in turn do a `createAuth` in the db handler (one of them will fail with a constraint violation)

**NOTE**: the only thing that doesn't make sense to me is that `create_player_on_auth_success` must be true in order for the above to happen simultaneously, but for that a default password has to be set, which wasn't the case when we tested this on YL :confused:

**Comment by sfan5** [comment]
The server thread processes packets from all clients in the same main loop with everything else:
https://github.com/minetest/minetest/blob/480d5f2d51ca8f7c4400b0918bb53b776e4ff440/src/server.cpp#L120-L122

Not to mention that entering the same Lua state from two threads would fail very quickly, Lua is neither thread-safe not multithreaded.

**Comment by 0siribix** [comment]
Since the auth is handled by lua, doesn't that effectively make this multi-threaded?

**Comment by fluxionary** [comment]
Pretty sure I triggered a similar issue using sqlite too, but one thing at a time. 

Unfortuantely, for some reason I now can't re-generate the original crash, just things similar to https://github.com/minetest/minetest/issues/11877#issuecomment-1107922494, and the occasional core dump. 

I will include postgres logs (including all executed statements), as well as @Oblomov's extra data in the minetest log, to which i appended a thread ID. 
```c++
#if DEBUG_DB
		warningstream << stmtName << " thread " << std::this_thread::get_id() << std::endl;
#endif
```
This is the result of 3 consecutive floods, #
[crashlogs.tar.gz](https://github.com/minetest/minetest/files/8557668/crashlogs.tar.gz)

One thing that is clear is that multiple threads are executing SQL statements, though it's hard to tell if they're doing it at the same time (or not).

**Comment by Festus1965** [comment]
I have it also,
stopped him/her with mod border = close for new gamer, then no crash

testing now small mod hint from LoineWolfHT
https://github.com/minetest/minetest/issues/12220#issuecomment-1108789409
is active in restarted server

**Comment by 0siribix** [comment]
[auth.txt](https://github.com/minetest/minetest/files/8559493/auth.txt)

I've edited my builtin/game/auth.lua to the following and seems to be working well so far

Limits total number of concurrent registrations to 5 (configurable)
Limits total number of registrations from one IP to 1 (configurable)
Time for IP slot to open after being closed is 5 seconds (configurable)
Protection against registering the same name multiple times

Caveat: If someone beings registration and cancels before completion, no one can register that name for 30s

This could be converted to an external mod but requires custom auth_handler which I need to test more

**Comment by fluxionary** [comment]
for people affected: postgres and sqlite are vulnerable, you can mitigate this issue by using the leveldb backend for the auth database

**Comment by Festus1965** [comment]
another question : 

what settings have all the server that was hit from this : 

I just realize that I had before, and now again: 
minetest.conf
**`disallow_empty_password = true`**

but during the attacks I am sure I had it open = false

normal I don't need gamer not even able to set a password, and lat them for other servers deal with them
but as of my not explained ban in forum I had more time and let them in,
and just see an Indonesian gamer 'gasan' need 4 times to check it out.

So we have following facts : 
* it depends on database (sqlite3 can't handle 2nd open)
* it can't pass default_password if sset
* a mod like border can block it
* the mod ipnames NOT, as the work does not reach so far, crash at auth

and so far three solutions yet:
* new auth.lua (0siribix)
* a prejoin.lua (LoneWoldHT)
* auth on LevelDB (nice, no crash, but massive login without use, and get maybe lag server)

**Comment by Festus1965** [comment]
I, Asia Thailand server, had about 14pm ICT a visitor ...
stinker, build a word N.... with alphabeth in nature - USA Cox Communications
but can't set an attack as normal happen with gamer like this one ...
just as of disallow_empty_password = true
all other my protections (default_password no, border = open), auth not new yet ...

but anyway : that have to be solved as we see options above.

**Comment by Festus1965** [comment]
also just came in : I also report IP Addresses to owner, here one answer,
what leads then to the mod vpnblocker:

<p>Dear Sir/Madam,</p>
      <p>Thank you for your message.</p>
      <p>The IP-address is assigned to a customer of Worldstream.<br>
        This customer is ProtonVPN. They offer VPN services to their end
        users.<br>
        We have informed of your message and asked them to look into
        this.<br>
        They might not able to fully resolve this issue.<br>
        Please feel free to block this IP-address if you continue to see
        abusive behavior.</p>
      <p>Kindly let us know if you require any more information or
        action from Worldstream.</p>
      <p>Met vriendelijke groet / Best regards,</p>
      

Thijs
--
Manager Trust & Safety
abuse@worldstream.com

```
Dear Sir/Madam,

Thank you for your message.

The IP-address is assigned to a customer of Worldstream.
This customer is ProtonVPN. They offer VPN services to their end users.
We have informed of your message and asked them to look into this.
They might not able to fully resolve this issue.
Please feel free to block this IP-address if you continue to see abusive behavior.

Kindly let us know if you require any more information or action from Worldstream.

Met vriendelijke groet / Best regards,
Thijs
Manager Trust & Safety
[abuse@worldstream.com](mailto:abuse@worldstream.com)
```

**Comment by fluxionary** [comment]
So the crash in the title can be fixed by changing the `auth_create` statement to something like this
```c++
       prepareStatement("auth_create",
               "INSERT INTO auth (name, password, last_login) VALUES ($1, $2, $3) "
                       "ON CONFLICT ON CONSTRAINT auth_pkey DO NOTHING "
                       "RETURNING id ");
```

I tried locking access to the db connection for the postgres auth connection, by adding this to the declaration of `AuthDatabasePostgreSQL`,
```c++
std::mutex db_mutex;
```

and adding 
```c++
MutexAutoLock lock(db_mutex);
```
to the top of `getAuth`, `saveAuth`, `createAuth`, `deleteAuth`, and `listNames`. However, I still get the same crash.

```
2022-04-26 15:54:54: ERROR[Main]: ServerError: AsyncErr: Lua: Runtime error from mod '' in callback getAuth(): PostgreSQL database error while checking results of execPrepared auth_read ERROR:  current transaction is aborted, commands ignored until end of transaction block
2022-04-26 15:54:54: ERROR[Main]: stack traceback:
2022-04-26 15:54:54: ERROR[Main]: 	[C]: in function 'read'
2022-04-26 15:54:54: ERROR[Main]: 	/opt/minetest-vanilla/bin/../builtin/game/auth.lua:14: in function </opt/minetest-vanilla/bin/../builtin/game/auth.lua:12>
```

Here's the postgres db code in case anyone wants to poke at it and see if I missed something. 
[dbpg.tar.gz](https://github.com/minetest/minetest/files/8567386/dbpg.tar.gz)

**Comment by Festus1965** [comment]
I just had a new attack:

Dannyboy : 111.216.47.x, Sony Network, = Japan, Osaka

as it is possible to change password to 'empty' he loged in, did some typical chat fun, and I see he relog with one false password attempt, so I guess already he changed =
he prepare to open this account for his script and
then he let it run

server stable, no crash
mod euban blocked anyway, as of banned using also a cheating client, he build himself up in air ... maybe multicraft2.0.0

so how to tell mt server, that empty password also is not allowed to set ?

**Comment by fluxionary** [comment]
> so how to tell mt server, that empty password also is not allowed to set ?

[`disallow_empty_password`](https://github.com/minetest/minetest/blob/480d5f2d51ca8f7c4400b0918bb53b776e4ff440/builtin/settingtypes.txt#L1163-L1164)

Also, this isn't the place to discuss the hacker spamming, it's more about the causes and solutions to the crash(es).

**Comment by Festus1965** [comment]
lol, are you joking me ?

the empty password is - so far I know - only entrance for the existing attack script to force the server with too much logins in short time and crash it !

My server is only safe yet, as I set this disallow true,
but now the script owner joins once, set the password for this account to empty,
and then starts the attack again to crash the server.

So beside this minetect.conf setting for a register,
IT DOES NOT work for a password change after register.

and so the hacker has another way to attack and crash server.

My look today as only, that I already ban him, and later add the IP range to ufw.
As he attacked later again, the faster option now is to use server command setpassword and change it, to block that 'attacking entrance' also.

Your answer show, that you didn't know this way even with set true, to get password empty again. I tested it, and then I wrote my post, to make sure every way to stop this crash attacks might be seen and used.
